### DISCUSSION TITLE
Neuron future

### DISCUSSION THREAD

**Discussion by Addonman** [brittyazel/Neuron#453] [Q&A]
Copy pasted it here as you wanted it here.

First of all thank you for all the work you have done to keep this addon alive up to this point. I have seen that you have put Neuron on hiatus for forsiable future but since then i have seen recent commits done to Neuron next . I just wanted to know what does that mean for future of Neuron ? Is perhaps Ylixir someone that might take over Neuron and continue to keep it alive since new commits that are merged are his?

**Comment by ylixir** [comment]
I want to see Neuron continue because it's unique and valuable to me. I don't want to make a major commitment or necessarily deal with managing an open source project.

These two wishes are in opposition to one another, so I have offered to maybe possibly think about considering lending a hand on Neuron and see which wins out.

If the project contnues, I think we are going to mothball the main branch and focus on the NeuronNext branch. Neither one of us has any interest in maintaining diverging codebases

**Comment by Addonman** [comment]
Thank you for that info. I just hope (if thats right word)  that you  come to your decision soon . I know its not an  easy decision because tbh just thinking of setting up all my alts bars again when i read that Neuron is going into hiatius gave me pause and i stoped playing for few days to see where addons (not just Neuron)  are going to end up in the end. So i can only imagine how time consuming is to maintain addon like Neuron especially  when blizz starts changing things. I am old user of Macaroon,Ion and now Neuron so i certinly hope someone would continue to maintain it so best wishes to anyone who hopefully takes this task be it you or someone else.

**Comment by brittyazel** [comment]
Like @ylixir said, the two diverging codebases of Neuron (Master branch) and NeuronNext was too much to handle. It hit a point where I had anxiety of developing much of anything on either branch due to having to duplicate the effort manually on the other. So there are likely going to be no more versions of Neuron (Master branch) with all new work going into to finishing NeuronNext and having it eventually supplant Master. It also gives me time for us to develop NeuronNext without having to also maintain a working version, which is a full time gig all on its own.

There are some major things to still be done though. We need to do lots of GUI work. We have some library porting that needs to be done. And Blizzard completely broke how SecureActionButtons work, so as a temporary patch we're using a workaround method for the buttons to even work. This change on Blizzards end has really, really f**cked us. They literally say in the base code:

> SecureActionButton_OnClick takes two extra custom arguments. Our usage of this template will always override this and supply the extra arguments, and for now AddOns won't be able to have isKeyPress or isSecureAction set.

and

> Why are we adding extra arguments, 'isKeyPress' and 'isSecureAction', to an _OnClick handler?
> We want to prevent mouse actions from triggering press-and-hold behavior for now, but we do want to allow AddOns to function as they did before. This is a problem since there's no difference between an AddOn's key press behavior and mouse click behavior. So if we don't know where this is coming from, it's from an AddOn and should be treated as a key press not a mouse press for 'useOnKeyDown' purposes.

Until they figure out how they want addons to interact with this new system. our hands are tied. That is unless @ylixir can figure out some freaking wizard magic to work around it without causing taint.

So yeah, in a nutshell. Putting Neuron on hiatus is allowing us to work on NeuronNext unimpeded. I now have some help, which is marvelous. And, most importantly, Blizzard broke everything. and we're still putting the pieces together.

**Reply by BujuArena** [reply]
This sounds like something that should be posted in https://github.com/Stanzilla/WoWUIBugs/issues/

**Reply by brittyazel** [reply]
Yup I reported it the day I found it:
https://github.com/Stanzilla/WoWUIBugs/issues/268

This is the most recent discussion:
https://github.com/Stanzilla/WoWUIBugs/issues/317

**Comment by Addonman** [comment]
Is there eta of when you think Neuronnext might be ready for public use? Or it might be still too much broken to be usefull? Or is it too far into future? Asking since prepatch event is coming soon and i would like to have little more info before i need to make a decision to setup my bars addon.

**Reply by brittyazel** [reply]
No ETAs.

**Reply by ylixir** [reply]
So, understanding that the `main` branch is alpha at best, and we don't need bug reports yet, we already know about enough bugs to keep us busy. Right now you can try out the latest code here:
https://github.com/brittyazel/Neuron/archive/refs/heads/main.zip

Make sure you put it in a folder named "Neuron" in your addons or it won't work. Also this only works in retail, classic takes some more processing.

You can actually download any version of the code, including stuff that is still not in `main` even. Just choose the proper branch on the left, and choose to download the zip on the right.
![image](https://user-images.githubusercontent.com/6045286/201484389-9b8200fc-6dfd-47aa-848d-ced7b55cbce1.png)


edited: NeuronNext->main

**Reply by ylixir** [reply]
Also, save your config! Data could easily be broken in development branches

You can back up the saved variables file, or from the experimental section of the neuron settings

**Comment by SparkusK** [comment]
Hey, I got NeuronNext up and running on Dragonflight but it's very buggy:
* `/neuron` commands don't work
* When `show grid` is enabled, buttons aren't shown
* When you increase amount of buttons on a bar, the counter doesn't visually increase, but the amount of buttons on the bar is increased
* When you increase amount of buttons on a bar and click on the bar again, it breaks the whole neuron ui, need to force a `/reload`
* Keybind labels disappear after a couple of reloads
* State modifiers + cooldown indicators seem to not be working as intended, if something's on cooldown and you swap between some `shift` state and back, the cooldown indicator remains floating in the background even for a button that isn't on cooldown

if we can get these (and maybe some others I haven't seen) fixed then things would be mostly functional for NeuronNext

**Reply by SparkusK** [reply]
I can maybe take a look into setting up a dev environment here, never done lua before but i'm sure it's interesting. Could someone maybe point me at some files to get started on fixing some of these things?

**Reply by ylixir** [reply]
That's great! I'm glad you are trying it out!

The `/neuron` commands are intentionally disabled. They are honestly pretty far down my priority list. I just assumed no one used them? If you are looking to help out and need some low hanging fruit to get started, this would be an excellent place to begin!

I can't reproduce any of the other issues.  Try to temporarily disable _ALL_ of your other addons and see if the issues persist? [Also make sure that you have downloaded the most current release from here?](https://github.com/brittyazel/Neuron/archive/refs/heads/main.zip) There were updates last night even!

If you still have troubles after those steps, then go ahead and open an issue or five in the issue section!

With regard to a development environment, I would recommend installing the addons bugsack and buggrabber at a bare minimum. They will allow you to keep an eye on lua errors without destroying your gaming experience. After that it all kind of depends on what text editor/ide you like, your comfort level with git and whether you like to live in windows/mac/linux.

To teach your text editor a little bit about the Warcraft runtime, you can access [language server library definitions](https://github.com/Ellypse/IntelliJ-IDEA-Lua-IDE-WoW-API/blob/master/GenerateFromAPIDocumentation.lua) here that work with [sumneko/lua-language-server](https://github.com/sumneko/lua-language-server) and [emmylua](https://emmylua.github.io/)

If you are using Windows, Mac, or native Linux, then you can just clone/unzip the Neuron code right into your addons folder. If you are using Windows Subsystem for Linux (WSL) then I wouldn't recommend that as the VM barrier makes for a very slow and painful experience. But I have a [branch with some quality of life features](https://github.com/ylixir/Neuron/blob/build-bones/Makefile) for keeping the code in WSL and mirroring it in your addons folder.

edited: update NeuronNext link to main

**Reply by SparkusK** [reply]
>The /neuron commands are intentionally disabled. They are honestly pretty far down my priority list. I just assumed no one used them? 

Fair, I thought it was broken, if it's intentionally disabled that's fine. I just used them to confirm the addon's presence and to get into the config, but the button works fine so np

>I can't reproduce any of the other issues.

Sweet! That means something else is broken on my side and I can get it working on local, that's actually relieving. Thanks!

>[Also make sure that you have downloaded the most current release from here?](https://github.com/brittyazel/Neuron/archive/refs/heads/NeuronNext.zip) There were updates last night even!

I actually DL'd your latest PR's payload branch, that's what's running here xD

Thanks for the tips on dev env, I mostly do Ruby on Linux in my work hours, will see if I can figure something out on Windows over time, Atom works here, might get that up and running, otherwise will just get VSCode or something haha.

**Reply by brittyazel** [reply]
Just a heads up, NeuronNext has been renamed to 'main' and is not the default branch.

**Reply by SparkusK** [reply]
Seems like I found the culprit - disabling `Masque` seems to have fixed a lot of abovementioned issues. Thanks for the feedback, folks. When I've got a bit more time I'll get a dev env going and see if I can make some contributions - always wanted to try Lua and project seems like fun!

**Reply by brittyazel** [reply]
Would be great to have you helping out!

I wonder if Masque updated their API again...

**Reply by ylixir** [reply]
https://github.com/brittyazel/Neuron/issues/467 could be related. fixed that last night

**Comment by Addonman** [comment]
Was playing with Neuronnext for last couple of days.Works failry ok for  what isnt out as alpha yet,but i come to obstacle to play and further testing. Its dragonflight. I tried few setting but i cant seem to get dragonrideing spells.Is there version that fixes that or you didnt come to that point or reworking Neuron?
Finally turned on Masque today as well and ofc that doesnt work as well.I just menitoned that since you talked about it in previous post.

**Reply by ylixir** [reply]
i honestly haven't even done dragonriding yet. what is the issue with those spells?

i did play the evoker, if it's like the evoker abilities, there should be an option in the blizzard config to do tap and release instead of hold

**Comment by Addonman** [comment]
I was on druid this time and you get like mount after this quest https://www.wowhead.com/quest=68795/dragonriding . When you mount it you are supposed to get bar with 2 spells + power bar with 4 eggs/jars of vigor  . According to this guide https://www.wowhead.com/guide/dragonriding/how-to-ride-your-dragon-dragonflight  we will have 4 spells and its like possess bar.  To finish quest you need to use those spells to get extra power at certain part of travel through rings ,but new bar didnt show up at all.

**Comment by brittyazel** [comment]
afaik we have exactly zero code to accommodate the dragonriding functionality, so I'm not surprised. I'm not sure if this is treated as a "state", in the same way as stances and the like are treated. If it is treated as a bar state, it may be simple enough to add it to our list of states and enable on Bar #1 by default.

**Comment by Addonman** [comment]
Guide says its a possession bar,but i have already turned on possesion,vehicle and overide option.It did nothing.
I turned also everything  on ,nothing happened. Only thing that did happen  me losing my buttons in all other forms but main. :( .
Luckily old WTF data saved the day.

**Reply by brittyazel** [reply]
This thread my be useful:
https://us.forums.blizzard.com/en/wow/t/custom-conditional-for-dragon-riding-bartender-showhide/1420191/15

> Expanding on what Elvie said, [bonusbar:11] might work. If his information is correct the dragon riding bar is bonusbar 11. That conditional exists.
> 
> Dragonriding = [bonusbar:11], not dragonriding = [nobonusbar:11]
> 
> I hope that helps.

**Reply by ylixir** [reply]
are bonus bars just stances? does turnimg on stances for the bar work? looks like soar is supposed to have it's own bar as well which is good news for my evi
okers awkward soar button bindings lol

**Comment by Addonman** [comment]
What is current status of Neuron? I havent seen any new fixes in couple of months or any news about it so just curious. 
Its in playable state in retail game and 10x for that but ofc there are few issues i would like to see fixed.
